Add full text search for queries based on the Postgres tsvector type.

jezdez opened the pull request:
This needs some more tests with a prod like database and probably another
look for someone with more experience with custom alembic migrations.

This is related to #1400 of course and uses [SQLAlchemy-searchable](https://sqlalchemy-searchable.readthedocs.io/en/latest/).

arikfr commented on migrations/versions/5ec5c84ba61e_.py:
```diff
@@ -0,0 +1,35 @@
+def downgrade():
+    conn = op.get_bind()
+
+    ss.drop_trigger(conn, 'article', 'search_vector')
```
`article`?

jezdez replied:
Yikes!

arikfr commented on redash/cli/database.py:
```diff
@@ -7,6 +7,8 @@
     """Create the database tables."""
     from redash.models import db
+    # make sure the additional search mappers are configured
+    db.configure_mappers()
```
What are the search mappers?

jezdez replied:
The comment is misleading, this basically makes sure that all mappers that have been registered with the SQLAlchemy ORM have been properly configured. Since SQLAlchemy-Searchable installs database triggers that have to act on when ORM model instances are created (e.g. to track updates), this call is needed.

Here's what the SA-Searchable docs say:

> It is very important here that you either access your searchable class or call configure_mappers before the creation of tables. SA-Searchable adds DDL listeners on the configuration phase of models.

Since this code happens in the phase in which tables are created, we can't really access the searchable class, but have to call `configure_mappers`. I'll update the comment to reflect this.

arikfr commented on redash/models.py:
```diff
@@ -915,27 +931,33 @@ def outdated_queries(cls):
-            where |= Query.id == term
-
-        where &= Query.is_archived == False
+    def search_queries(cls, term, group_ids, include_drafts=False, limit=20):
```
Why append `_queries`? It's a method on the `Query` class isn't it obvious that it searches for queries? And in the handlers code you still call `.search`...

jezdez replied:
Ugh, yeah, this is just a leftover from figuring out how to use the `search` method that the `SearchQueryMixin` adds to the `query_class` of the `Query` model. Will fix.

arikfr commented on redash/models.py:
```diff
@@ -28,10 +28,12 @@
 from sqlalchemy.types import TypeDecorator
 from functools import reduce
+from sqlalchemy_searchable import SearchQueryMixin, make_searchable, search
+from sqlalchemy_utils.types import TSVectorType
```
Not that I mind much about adding `sqlalchemy_utils` as a dependency, but doesn't `sqlalchemy_searchable` provide something for this?

jezdez replied:
Yeah, the [SQLAlchemy-Utils](https://github.com/kvesteri/sqlalchemy-utils) package is from the same author as [SQLAlchemy-Searchable](https://github.com/kvesteri/sqlalchemy-searchable) and contains a whole bunch of things.

The code behind [TSVectorType](https://github.com/kvesteri/sqlalchemy-utils/blob/master/sqlalchemy_utils/types/ts_vector.py) is trivial relatively speaking, so I guess we could also just ship our own. The only risk I see with this is that we could trade one dependency less for having to keep an eye on changes in SQLAlchemy-Searchable that require changes in the TSVectorType class as well.

On the other hand, although quite out of scope of this PR, is the question if there are things in SQLAlchemy-Utils that may be useful for Redash in general, given its swiss-army-knife-ness. I'm fine with whatever way you want to go with this.

arikfr replied:
👍 ok, let's use it. No need to find more places to use it, we can gradually introduce it where applicable.

arikfr commented on migrations/versions/5ec5c84ba61e_.py:
```diff
@@ -0,0 +1,35 @@
+                                       nullable=True))
+
+    ss.sync_trigger(conn, 'queries', 'search_vector',
+                    ['name', 'description', 'query'])
```
Will this also update all existing queries?

jezdez replied:
Yes, looking at the [docs](https://sqlalchemy-searchable.readthedocs.io/en/latest/alembic_migrations.html#sqlalchemy_searchable.sync_trigger) it says: "Updates all rows for given search vector by running a column=column update query for given table."

I've also manually added a bunch of queries against Redash's own Postgres database in the master branch, then switch to this branch, did the Alembic upgrade and then was able to use the search feature right away.

arikfr commented on migrations/versions/5ec5c84ba61e_.py:
```diff
@@ -0,0 +1,35 @@
+    conn = op.get_bind()
+
+    ss.drop_trigger(conn, 'queries', 'search_vector')
+    op.drop_index('ix_queries_search_vector', table_name='queries')
```
This is a byproduct of `ss.sync_trigger`?

jezdez replied:
Yep, although weirdly it raised a ProgrammingError exception when I tried roll it back, will look into this.

jezdez replied:
This was an oversight when setting up SA-Searchable. The migration that I fixed in https://github.com/getredash/redash/pull/2041/commits/eb6966fa83d2b44978bf5b49db0df052d11c8170 now reflects adding an index for the search vector column. Another call to `db.configure_mappers()` was needed to have the mapper processor of SA-Searchable to correctly add the index.

arikfr replied:
👍 

> Another call to db.configure_mappers() was needed

It seems that it's at the top level of `redash.models`. So I guess the one in `redash.cli.database` is redundant?

jezdez replied:
True indeed, good catch, removed it in https://github.com/getredash/redash/pull/2041/commits/36aeef776d610a266d1c9dfa166be10bb5d1726a.

jezdez commented:
@arikfr Just added a bunch of tests for the more advanced search queries as described on https://sqlalchemy-searchable.readthedocs.io/en/latest/search_query_parser.html.

arikfr commented:
@jezdez this is finalyl ready to be merged, but I think I saw some fixes you did to this code in Mozilla's fork. Are they present here already or this need an update?

Thanks.

jezdez commented:
@arikfr They aren't in this PR yet, I'll get them in today and let you know. Thanks for reaching out again!

arikfr commented:
Super, thanks!

jezdez commented:
@arikfr Okay, I've updated the PR with the recent things in our fork and updated the design to match the query-list and my-queries page as well. Does this work?

<img width="984" alt="screen-shot-2018-01-08-22-48-58" src="https://user-images.githubusercontent.com/1610/34694104-2e0f3c3c-f4c6-11e7-9538-5665731a3f8a.png">

jezdez commented:
@arikfr BTW, apologies for taking so long :(

jezdez commented:
@arikfr Any idea how we can retrigger the code climate check? It seems to be stuck.

arikfr commented:
Thanks ! 👍
